File: problems/p698/Solution.py
class Solution:
    def canPartitionKSubsets(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: bool
        """
        if not nums:
            return False

        if k == 1:
            return True

        if len(nums) < k:
            return False

        total = sum(nums)
        if total % k != 0:
            return False

        target = total // k

        sorted_nums = sorted(nums)
        buckets = [0 for _ in range(k)]

        return self.dfs_util(target, buckets, sorted_nums, len(nums) - 1)

    def dfs_util(self, target, buckets, nums, idx):
        if idx == -1:
            for bu in buckets:
                if bu != target:
                    return False
            return True

        num = nums[idx]
        for i in range(len(buckets)):
            if buckets[i] + num > target:
                continue

            buckets[i] += num
            if self.dfs_util(target, buckets, nums, idx - 1):
                return True
            buckets[i] -= num

        return False

File: problems/p698/Solution.java
import java.util.Arrays;

public class Solution {
    public boolean canPartitionKSubsets(int[] nums, int k) {
        if(nums == null || nums.length == 0) 
            return false;
        
        if(k > nums.length) 
            return false;
        
        int total = 0;
        for(int x : nums) {
            total += x;
        }
        
        if (total % k != 0)
            return false;
        
        Arrays.sort(nums);
        int target = total / k;
        int[] bucket = new int[k];
        return dfsUtil(bucket, target, nums, nums.length - 1);
    }
    
    private boolean dfsUtil(int[] bucket, int target,  int[] nums, int idx) {
        if (idx == -1) {
            for(int x : bucket) {
                if(x != target) {
                    return false;
                }
            }    
            return true;
        }
        
        int num = nums[idx];
        for(int i = 0; i < bucket.length; i++) {
            if (bucket[i] + num > target) 
                continue;
            bucket[i] += num;
            
            if (dfsUtil(bucket, target, nums, idx - 1)) {
                return true;
            }

            bucket[i] -= num;
        }
        
        return false;
    }
}
